File: python/net/recv.py
import logging
from socket import getaddrinfo
from socket import socket

import fire
from common.logSetup import logSetup
from common.modeHandle import modeHandle
from enums import Mode

# Main #################################################################################


def main(host: str, port: int, mode: Mode = Mode.TCP4, buf: int = 1024):
    logSetup()
    logging.info(
        "Entering recv.main with host = {}, port = {}, mode = {}.".format(
            host, port, mode
        )
    )

    # Process Arguments ================================================================

    mode, aFam, sKind = modeHandle(mode)

    # Get IP ===========================================================================

    aInfo = list(set(getaddrinfo(host, port, aFam, sKind)))[0]
    ip = aInfo[4][0]
    logging.info("Using host ip: '{}'.".format(ip))

    # Do socket stuff ==================================================================

    with socket(aFam, sKind) as s:
        logging.debug("Trying to bind...")
        s.bind((ip, port))
        logging.debug("Bound.")

        if mode == Mode.TCP4 or mode == Mode.TCP6:
            logging.debug("Trying to listen...")
            s.listen()
            logging.debug("Listening.")

            conn, addr = s.accept()
            with conn:
                logging.info("Got connection from '{}'.".format(addr))
                while True:
                    data = conn.recv(buf)  # This call blocks until it gets data
                    if not data:
                        logging.error("Got no data during recieve.")
                        break
                    logging.info("Recieved data: '{}' from '{}'.".format(data, addr))
        elif mode == Mode.UDP4 or mode == Mode.UDP6:
            logging.info("Reading UDP...")
            while True:
                data, addr = s.recvfrom(buf)
                if not data:
                    logging.error("Got no data from UDP.")
                    break
                logging.info("Recieved data: '{}' from '{}'.".format(data, addr))
        else:
            raise NotImplementedError()


# Run ##################################################################################

if __name__ == "__main__":
    fire.Fire(main)

File: rust/net/src/recv.rs
mod enums;

use crate::enums::Mode;
use clap::Parser;
use log::{debug, info};
use std::io::Read;
use std::net::TcpListener;

// CLI /////////////////////////////////////////////////////////////////////////////////////////////

#[derive(Parser, Debug)]
#[command(author,version,about,long_about=None)]
struct Args {
  #[arg(long)]
  host: String,

  #[arg(short, long)]
  port: u16,

  #[arg(value_enum, short, long, default_value_t=Mode::TCP)]
  mode: Mode,
}

// Primary Function ////////////////////////////////////////////////////////////////////////////////

#[derive(Debug)]
pub enum RecvError {
  FailedToBind,
  FailedToAccept,
  FailedToRead,
  FailedUTF8Conversion,
  NotImplemented,
}

pub fn recv(host: String, port: u16, mode: Mode) -> std::result::Result<Vec<u8>, RecvError> {
  match mode {
    Mode::TCP => {
      debug!("Trying to bind to {host}:{port}.");
      let listener_result = TcpListener::bind(format!("{}:{}", host, port));
      match listener_result {
        Ok(listener) => {
          debug!("Waiting to accept connections...");
          let accept_result = listener.accept();
          match accept_result {
            Ok((mut stream, addr)) => {
              debug!("Trying to read from tcp stream...");
              let mut buf = [0u8; 1024];
              let read_result = stream.read(&mut buf);
              match read_result {
                Ok(_) => Ok(buf.to_vec()),
                Err(_) => Err(RecvError::FailedToRead),
              }
            }
            Err(_) => return Err(RecvError::FailedToAccept),
          }
        }
        Err(_) => return Err(RecvError::FailedToBind),
      }
    }
    Mode::UDP => {
      return Err(RecvError::NotImplemented);
    }
  }
}

// Main ////////////////////////////////////////////////////////////////////////////////////////////

fn main() -> std::result::Result<(), RecvError> {
  env_logger::init();

  info!("Parsing arguments...");
  let args = Args::parse();

  let recvResult = recv(args.host, args.port, args.mode);
  match recvResult {
    Ok(msg) => {
      let string = match std::str::from_utf8(&msg) {
        Ok(m) => m,
        Err(_) => return Err(RecvError::FailedUTF8Conversion),
      };
      println!("Recieved message {}", string);
      Ok(())
    }
    Err(e) => Err(e),
  }
}
